Kobo: Refactor various aspects of the Kaleido/MTK support

NiLuJe opened the pull request:
NOTE: Initial, deprecated, comment about the original approach that tried to handle this in UIManager:

This... is not pretty, and was mostly not a problem previously, because it mostly did not affect flashing updates, which means that, even if we were to request spuriously full-screen refreshes, they wouldn't do anything visible, because most waveform modes do not visibly affect unmodified pixels.

~~I'm hoping to be able to shove it under the rug once more once the kernel sources come out, because I suspect we don't actually need dedicated public color/colortext modes, and can instead just implement it silently behind the scenes in fb_mxcfb based on the dither flag.~~

EDIT: Turns out most of my issues with my initial test of the above were due to stupid bugs, so, done. Meaning this now depends on https://github.com/koreader/koreader-base/pull/1865

Fix #12218

----

Actual changelog now that things are simpler (via the magic of deleting code \o/):

* UIManager: Let the fb backend deal with Kaleido wfm promotion. This fixes a number of quirks that poisoned the refresh queue with spurious full-screen refreshes. See https://github.com/koreader/koreader-base/pull/1865 for more details.
* This also means we now disable Kaleido waveform modes when color rendering is disabled (remember to trash your thumbnail cache if you don't want to mix color w/ grayscale thumbnails, though).
* UIManager: Merge refreshes that share an edge, because that was driving me nuts (and would have most likely been merged by the kernel anyway). A perfect test-case is the FM, which trips two separate refreshes because of its title bar.
* ReaderFlipping: Use sensible dimensions, so that we only refresh the icon's region.
* ReaderBookmark: Only refresh the dogear instead of the whole page when toggling bookmarks.
* NetworkSetting: Make it a real boy, so it consistently refreshes properly on dismiss instead of relying on UIManager saving the day.
* Kobo: Aggressively prevent *both* suspend & standby while MTK devices are plugged-in, as both will horribly implode the kernel (we previously only prevent standby while charging).
* Kobo: Switch to 8bpp on B&W MTK devices (or when color rendering is disabled on Kaleido panels).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/12221)
<!-- Reviewable:end -->

NiLuJe committed:
UIManager: Mangle color handling so as not to always imply a full-screen
region...

Fix #12218

My hope is to be able to clear that up once the kernel sources clarify
that GCC16 is just GC16 + HWTCON_FLAG_CFA_MODE_G1 and
GLRC16 is just REAGL + HWTCON_FLAG_CFA_MODE_G1;
at which point we can just drop the entire color/colortext mess from
frontend, and just implement that based on the dither flag in the fb
backend...

NiLuJe committed:
Minor review pass

NiLuJe committed:
More comment tweaks

NiLuJe committed:
Drop debugging

NiLuJe committed:
Don't use Kaleido wfm if color is disabled

NiLuJe committed:
Not actually the lowest, but see dither above for why we don't use fast
;p.

NiLuJe commented:
Not that a much simpler test-case than the stuff mentioned in the linked issue is to simply dismiss the status bar ;).

Similarly, it definitely affected stuff like the CRe reloading icons, possibly even the loading progress bar I guess?

NiLuJe committed:
Geom: Add open range variants of intersectWith and use them in
UIManager's refresh collision test

This had been driving me nuts forever.

FM refreshes with a title-bar update are now properly coalesced into a
single refresh.

NiLuJe committed:
Explain why we sorta relied on full-screen regions before

That was mainly to make it stick via the intersect check,
knowing that larger-than-necessary refresh regions are generally
harmless (and invisible), except for flashing refreshes.

Which is where our troubles began, because color refreshes are always
flashing ;o).

NiLuJe committed:
Oh, and it being empty is sortof relevant, too.

NiLuJe committed:
Moar comments

NiLuJe committed:
ReaderFlipping: use the actual widget dimension for both sides

This made absolutely no sense, as scaleBySize is orientation-agnostic.

This lead to refreshes that were unnecessarily as wide as the screen,
instead of constrained to the actual icon.

NiLuJe committed:
Welp, can't fixup ReaderDogear the same way, unfortunately.

Anyway, readerbookmark just does a full-screen ui refresh anyway,
so it's mostly harmless besides that eye-gougingly painful resetLayout
hackery.

NiLuJe commented:
Toggling a bookmark still flashes, because https://github.com/koreader/koreader/blob/761cf1822258da62b5c72063f2b8b12d348b5cb4/frontend/apps/reader/modules/readerbookmark.lua#L315

I'm unfamiliar with the whole bookmark icons and how/where they're rendered, but dogear is already annoying enough that I don't think we have a much better way to deal with this without basically rewriting the both of them to be less annoying to catch as partial refreshes.

NiLuJe committed:
ReaderBookmark: Only refresh the dogear region in ToggleBookmark

hius07 commented:
Toggling a bookmark also updates the footer (Number of bookmarks indicator).

NiLuJe commented:
There, clobbered ReaderBookmark into submission.

----

> Toggling a bookmark also updates the footer (Number of bookmarks indicator).

There's already an onUpdateFooter call that handles that ;). (a bit optimistcally, mind you, as it refreshes even when you don't actually display the bookmark count).

NiLuJe committed:
ReaderBookmark: refresh the dogear first, so it's the only one that
inherits ReaderUI refresh quirks.

Which makes more sense than punting it off on poor ReaderFooter ;).
And, more importantly, makes sure it doesn't risk inheriting a flash on
color panels, so the driver can happily optimize the useless refresh
out.

NiLuJe commented:
> (a bit optimistcally, mind you, as it refreshes even when you don't actually display the bookmark count

Eeh, we don't really have a sane way to tell from outside ReaderFooter. I'll let the kernel optimize that one out.

NiLuJe committed:
More comments

NiLuJe commented:
> I'm hoping to be able to shove it under the rug once more once the kernel sources come out, because I suspect we don't actually need dedicated public color/colortext modes, and can instead just implement it silently behind the scenes in fb_mxcfb based on the dither flag.

Tried it, turns out I kinda like the granularity afforded by the current approach. :D.

NiLuJe committed:
Let the caller handle the sentinel region thingy

We don't necessarily want to break the current contract that nil ==
fullscreen, and we only have a single caller doing this anyway.

NiLuJe committed:
Match what happens on !isColorSCreens devices when !isColorEnabled

NiLuJe committed:
Drop debug

NiLuJe committed:
NetworkSetting: Explicitly request a regional refresh on close

Don't rely on UIManager quirks to do a full-screen refresh

NiLuJe committed:
Random comment reflow

NiLuJe committed:
ReaderView: Only flag CRe pages w/ actual image content as dithered

NiLuJe committed:
Kobo: Drive-by PM fixup on MTK.

Turns out it horribly crashes when attempting *both* standby and suspend
while chraging (and I suspect while charged, as long as it's plugged in).

So, take the bull by the horn, and nerf the imps directly so we catch
any and all sources of such requests...

Because I already killed it three times while testing this PR -_-".

NiLuJe committed:
Rely on automatic kaleido promotion in the backend

NiLuJe committed:
We can simplify that

The multiple checks involving checking for the dithered flag in the
widget stack do the job just fine, no need to special-case that further.

NiLuJe committed:
And now that the bugs are squashged, we can drop the night's experiment

NiLuJe committed:
And drop the explicit color/colortext refresh modes

The backend does the magic automatically.

NiLuJe committed:
And drop debug ;)

NiLuJe commented:
> Tried it, turns out I kinda like the granularity afforded by the current approach. :D.

Granularity issue turned out to be a bug, so that works just fine, and greatly simplifies this mess, so I'm happy.

NiLuJe committed:
Review pass, that comment is now mostly outdated ;).

poire-z commented on frontend/ui/geometry.lua:
```diff
@@ -208,6 +208,32 @@ function Geom:intersectWith(rect_b)
+
+@tparam Geom rect_b
+]]
+function Geom:notOpenIntersectWith(rect_b)
```
What does "open" means in this context ? If it is about the edges being shared, "open" does not intuitively relate to that for me :)

NiLuJe replied:
Half-open range, e.g., `[x, x + w)` instead of [x, x+ w]`.

poire-z commented on frontend/apps/reader/modules/readerflipping.lua:
```diff
@@ -40,18 +40,13 @@ function ReaderFlipping:init()
         alpha = true,
     }
     self[1] = LeftContainer:new{
-        dimen = Geom:new{w = Screen:getWidth(), h = self.flipping_widget:getSize().h},
+        dimen = Geom:new{w = self.flipping_widget:getSize().w, h = self.flipping_widget:getSize().h},
```
I guess this LeftContainer is a no-op/null-container then :) and we could have `self[1] = self.flipping_widget`
But ok for keeping it as-is and not change anything else that works.

NiLuJe replied:
Huh, yeah, looks like it.

NiLuJe replied:
Oh, wait, it might be useful to handle mirroring in RTL mode?

poire-z replied:
Oh, possibly indeed :)
So, may be your change of reducing its width from the screen width to the icon size is wrong ?

NiLuJe replied:
Oh, dammit, I was just thinking this might at least break the *refreshes* in RTL, but you're right, it probably breaks positioning, too :/.

Welp, I desperately need a nap, so, more on that later ;).

NiLuJe replied:
There, fixed it. And did something slightly more elegant for the refresh region (i.e., we can just use the IconWidget's dimen object).

poire-z reviewed (approved):
Looks ok and trusting you.

NiLuJe committed:
Simplify the charging/charged suspend/standby check

isCharging is still true when isCharged flips to true.

And, yes, kernel still implode when fully charged, just
quadruple-checked, for science!

NiLuJe committed:
Simplify the ReaderBookmark/ReaderDogear refresh region shenanigans

We can just use IconWidget's dimen, which is updated on paint.

Drive-by ListMenuItem tweak to not share dimen objects, because we've
updated LeftContainer to store its own position lke RightContainer.

poire-z commented on frontend/apps/reader/modules/readerbookmark.lua:
```diff
@@ -310,9 +310,15 @@ end
+        self.view.dogear:getRefreshRegion()
+    end)
+    -- And ask for a footer refresh, in case we have bookmark_count enabled.
+    -- Assuming the footer is visible, it'll request a refresh regardless, but the EPDC should optimize it out if no content actually changed.
```
Is that true for older boards ? (ie. my GloHD :))

NiLuJe replied:
Even if the driver doesn't actually optimize it out, even GC16 won't visually affect unchanged pixels, and that's been the case forever ;).

NiLuJe reviewed (commented):
> Reviewed 1 of 4 files at r2, 1 of 2 files at r3, 6 of 7 files at r6, 1 of 1 files at r8, 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12221)* status: all files reviewed, 3 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

NiLuJe committed:
Amend that comment, we've tamed the beast and should now be able to
switch the Clara B&W to actual B&W!

NiLuJe committed:
Hide the "no 8bpp" menu entry on color devices

Because duh'

NiLuJe committed:
Now that we can do that, swap to 8bpp when color is disabled on Kobo

Gotta go fast! (but still suffer from the physical aspects of the panel
sandwich :/).

NiLuJe committed:
Bump base

https://github.com/koreader/koreader-base/pull/1860
https://github.com/koreader/koreader-base/pull/1862
https://github.com/koreader/koreader-base/pull/1863 (fix https://github.com/koreader/koreader/issues/12207)
https://github.com/koreader/koreader-base/pull/1865

NiLuJe reviewed (commented):
> Reviewed 5 of 5 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12221)* status: all files reviewed, 3 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)



<!-- Sent from Reviewable.io -->

KraXen72 commented:
do these also improve things on android-based kaleido tablets like the onyx boox go color 7? or is this kobo specific?

NiLuJe commented:
Nope, Kobo only. There's (mostly) nothing we can do about Android. Cue my usual "anybody using an Android eInk tablet is some kind of masochist" comment ;).

KraXen72 commented:
> anybody using an Android eInk tablet is some kind of masochist
  
why though? and why couldn't these improvements, if even necessary, be applied to android eink tablets?

i have a pocketboox touch lux 4 but i just bought a boox go color 7 because it's much easier to get manga on it right from the device (either using mihon/tachiyomi or an android torrent client)

NiLuJe commented:
Because we have nowhere *near* this level of low-level control on Android (or PocketBook, for that matter).

KraXen72 commented:
Hello. sorry to necro this, I just got a Kindle Paperwhite 2024. How's the e-ink control story there? Thanks.

NiLuJe commented:
Pretty good, why?

KraXen72 commented:
Wanted to know if the e-ink supports partial updates when inside of koreader, and also, if i can expect it to be meaningfully faster in page-turning compared to my old onyx boox page, which is an android-based e-reader. and any other meaningful differences between those two in terms of koreader, i fyou can think of some off the top of your head.
Thanks!

NiLuJe commented:
Yes, to all of those. Android is terrible for eInk.

KraXen72 commented:
Alright, Thank you! and have a nice day :)))
